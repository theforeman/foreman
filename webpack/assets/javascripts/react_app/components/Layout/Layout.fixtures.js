const mockOnClick = jest.fn();

const PFitems = [
  {
    title: 'Monitor',
    initialActive: true,
    iconClass: 'fa fa-tachometer',
    subItems: subItemsA,
    href: '/a',
  },
  {
    title: 'Hosts',
    initialActive: false,
    iconClass: 'fa fa-server',
    subItems: subItemsB,
    href: '/b',
  },
  {
    title: 'Configure',
    initialActive: false,
    iconClass: 'fa fa-wrench',
    subItems: subItemsC,
    href: '/c',
  },
];

const subItemsA = [
  {
    title: 'Aa',
    isDivider: false,
    onClick: mockOnClick,
  },
  {
    title: 'Cc',
    isDivider: false,
    onClick: mockOnClick,
  },
];
const subItemsB = [
  {
    title: 'Dd',
    isDivider: false,
    onClick: mockOnClick,
  },
  {
    title: 'Ff',
    isDivider: false,
    onClick: mockOnClick,
  },
];
const subItemsC = [
  {
    title: 'Gg',
    isDivider: false,
    onClick: mockOnClick,
  },
  {
    title: 'Ii',
    isDivider: false,
    onClick: mockOnClick,
  },
];

// Server Hash Data
const monitorChildren = [
  {
    type: 'item',
    name: 'Dashboard',
    url: '/',
  },
  {
    type: 'item',
    name: 'Facts',
    url: '/fact_values',
  },
];

const hostsChildren = [
  {
    type: 'item',
    name: 'All Hosts',
    url: '/hosts/new',
  },
  {
    type: 'item',
    name: 'Architectures',
    url: '/architectures',
  },
];

const userChildren = [
  {
    type: 'item',
    name: 'Environments',
    url: '/environments',
  },
  {
    type: 'item',
    name: 'Architectures',
    url: '/architectures',
  },
];
const infrastructureChildren = [
  {
    type: 'item',
    name: 'Domains',
    url: '/domains',
  },
  {
    type: 'item',
    name: 'Realms',
    url: '/realms',
  },
];

const hashItemsA = [
  {
    type: 'sub_menu',
    name: 'Monitor',
    icon: 'fa fa-tachometer',
    children: monitorChildren,
    active: false,
  },
  {
    type: 'sub_menu',
    name: 'Hosts',
    icon: 'fa fa-server',
    children: hostsChildren,
    active: false,
  },
];

const hashItemsB = [
  {
    type: 'sub_menu',
    name: 'User',
    icon: 'fa fa-wrench',
    children: userChildren,
    active: false,
  },
  {
    type: 'sub_menu',
    name: 'Infrastructure',
    icon: 'pficon pficon-network',
    children: infrastructureChildren,
    active: false,
  },
];

export const serverItems = [...hashItemsA, ...hashItemsB];

const logo =
  '/assets/header_logo-c9614c16f2ee399ae9cb7f36ec94b9a26bf8cf9eabaa7fe6099bf80d1f7940db.svg';
const user = {
  current_user: {
    user: {
      id: 4,
      login: 'admin',
      firstname: 'Admin',
      lastname: 'User',
    },
  },
  user_dropdown: [
    {
      type: 'sub_menu',
      name: 'User',
      icon: 'fa fa-user',
      children: subItemsA,
    },
  ],
};

const organizations = {
  current_org: 'org1',
  available_organizations: [
    { id: 1, title: 'org1', href: '/organizations/1-org1/select' },
    { id: 2, title: 'org2', href: '/organizations/2-org2/select' },
  ],
};

const locations = {
  current_location: 'london',
  available_locations: [
    { id: 1, title: 'yaml', href: '/locations/1-yaml/select' },
    { id: 2, title: 'london', href: '/locations/2-london/select' },
    { id: 3, title: 'norway', href: '/locations/3-norway/select' },
  ],
};

const serverUser = {
  current_user: {
    user: {
      firstname: 'G',
      lastname: 'L',
    },
  },
  user_dropdown: [
    {
      children: [
        {
          type: 'item',
          url: '/',
          name: 'My Account',
        },
        { type: 'divider' },
      ],
    },
  ],
};

export const layoutMock = {
  items: PFitems,
  data: {
    menu: [...hashItemsA, ...hashItemsB],
    locations,
    orgs: organizations,
    root: '/',
    logo,
    notification_url: '/notification_recipients',
    taxonomies: { locations: true, organizations: true },
    user,
  },
};

export const noItemsMock = {
  ...layoutMock,
  items: [],
};

export const hasTaxonomiesMock = {
  ...layoutMock,
  currentLocation: 'london',
  currentOrganization: 'org1',
};

export const userDropdownProps = {
  user: serverUser,
  notification_url: '/',
};
