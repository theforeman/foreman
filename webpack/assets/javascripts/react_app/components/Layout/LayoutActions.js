import {
  LAYOUT_INITIALIZE,
  LAYOUT_SHOW_LOADING,
  LAYOUT_HIDE_LOADING,
  LAYOUT_CHANGE_LOCATION,
  LAYOUT_CHANGE_ACTIVE,
  LAYOUT_CHANGE_ORG,
  LAYOUT_EXPAND,
  LAYOUT_COLLAPSE,
} from './LayoutConstants';

export const initializeLayout = ({
  items,
  activeMenu,
  isCollapsed,
  organization,
  location,
}) => ({
  type: LAYOUT_INITIALIZE,
  payload: {
    items,
    activeMenu,
    isCollapsed,
    organization,
    location,
  },
});

export const showLoading = () => ({
  type: LAYOUT_SHOW_LOADING,
});

export const hideLoading = () => ({
  type: LAYOUT_HIDE_LOADING,
});

export const changeActiveMenu = ({ title }) => ({
  type: LAYOUT_CHANGE_ACTIVE,
  payload: {
    activeMenu: title,
  },
});

export const changeOrganization = organization => ({
  type: LAYOUT_CHANGE_ORG,
  payload: { organization },
});

export const changeLocation = location => ({
  type: LAYOUT_CHANGE_LOCATION,
  payload: {
    location,
  },
});

export const expandLayoutMenus = () => ({
  type: LAYOUT_EXPAND,
});

export const collapseLayoutMenus = () => ({
  type: LAYOUT_COLLAPSE,
});
