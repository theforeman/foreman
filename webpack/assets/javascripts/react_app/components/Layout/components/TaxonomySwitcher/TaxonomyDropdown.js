import React, { useState, useEffect, useCallback } from 'react';
import PropTypes from 'prop-types';
import { Button, Grid, GridItem, Icon } from '@patternfly/react-core';
import {
  ContextSelector,
  ContextSelectorItem,
  ContextSelectorFooter,
} from '@patternfly/react-core/deprecated';
import { CheckIcon, GlobeIcon, BuildingIcon } from '@patternfly/react-icons';
import { foremanUrl } from '../../../../common/helpers';
import { translate as __ } from '../../../../common/I18n';
import './TaxonomyDropdown.scss';

const TaxonomyDropdown = ({ taxonomyType, currentTaxonomy, taxonomies }) => {
  const id = `${taxonomyType}-dropdown`;
  const anyTaxonomyURL = foremanUrl(`/${taxonomyType}s/clear`);
  const manageTaxonomyURL = foremanUrl(`/${taxonomyType}s`);
  const anyTaxonomyText =
    taxonomyType === 'organization'
      ? __('Any organization')
      : __('Any location');

  const [searchValue, setSearchValue] = useState('');
  const [isOpen, setIsOpen] = useState(false);
  const [filteredItems, setFilteredItems] = useState(taxonomies);

  const onSearchButtonClick = useCallback(() => {
    const filtered =
      searchValue === ''
        ? taxonomies
        : taxonomies.filter(item =>
            item.title.toLowerCase().includes(searchValue.toLowerCase())
          );
    setFilteredItems(filtered || []);
  }, [searchValue, taxonomies]);

  useEffect(() => {
    onSearchButtonClick();
  }, [searchValue, onSearchButtonClick]);

  const onToggle = (event, newIsOpen) => {
    setIsOpen(newIsOpen);
  };
  const onSelect = () => {
    setIsOpen(!isOpen);
  };
  const onSearchInputChange = (value, event) => {
    setSearchValue(event.target.value);
  };

  const selectedIcon = (
    <Icon size="sm">
      <CheckIcon className="current-taxonomy-v" />
    </Icon>
  );
  const anyIcon =
    taxonomyType === 'organization' ? (
      <BuildingIcon style={{ marginRight: '5px', marginTop: '3px' }} />
    ) : (
      <GlobeIcon style={{ marginRight: '5px', marginTop: '3px' }} />
    );

  const anyTaxonomyItem = (
    <ContextSelectorItem
      key={0}
      className={`${taxonomyType}s_clear`}
      onClick={() => {
        window.location.assign(anyTaxonomyURL);
      }}
      isDisabled={!currentTaxonomy}
    >
      <Grid hasGutter>
        <GridItem span={1}>{anyIcon}</GridItem>
        <GridItem span={9} style={{ textAlign: 'left' }}>
          {anyTaxonomyText}
        </GridItem>
        <GridItem span={2}>{!currentTaxonomy && selectedIcon}</GridItem>
      </Grid>
    </ContextSelectorItem>
  );
  const footer = (
    <ContextSelectorFooter>
      <Button
        ouiaId={`manage-taxonomy-button-${taxonomyType}`}
        size="sm"
        component="a"
        className={taxonomyType}
        variant="secondary"
        href={manageTaxonomyURL}
      >
        {taxonomyType === 'organization'
          ? __('Manage Organizations')
          : __('Manage Locations')}
      </Button>
    </ContextSelectorFooter>
  );
  return (
    <ContextSelector
      ouiaId={`taxonomy-context-selector-${taxonomyType}`}
      id={id}
      toggleText={
        currentTaxonomy || (
          <>
            {anyIcon}
            {anyTaxonomyText}
          </>
        )
      }
      onSearchInputChange={onSearchInputChange}
      isOpen={isOpen}
      searchInputValue={searchValue}
      onToggle={onToggle}
      onSelect={onSelect}
      onSearchButtonClick={onSearchButtonClick}
      screenReaderLabel="Selected Taxonomy:"
      className="context-selector"
      footer={footer}
    >
      {anyTaxonomyItem}
      {filteredItems.map(({ title, href }, i) => (
        <ContextSelectorItem
          key={i + 1}
          id={`select_taxonomy_${title}`}
          className={`${taxonomyType}_menuitem`}
          onClick={() => {
            if (href) {
              window.location.assign(href);
            }
          }}
          isDisabled={title === currentTaxonomy}
        >
          <Grid hasGutter>
            <GridItem span={10} style={{ textAlign: 'left' }}>
              {title}
            </GridItem>
            <GridItem span={1}>
              {title === currentTaxonomy && selectedIcon}
            </GridItem>
          </Grid>
        </ContextSelectorItem>
      ))}
    </ContextSelector>
  );
};

TaxonomyDropdown.propTypes = {
  taxonomyType: PropTypes.oneOf(['organization', 'location']).isRequired,
  currentTaxonomy: PropTypes.string,
  taxonomies: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number.isRequired,
      title: PropTypes.string,
      href: PropTypes.string.isRequired,
    })
  ).isRequired,
};

TaxonomyDropdown.defaultProps = {
  currentTaxonomy: undefined,
};

export default TaxonomyDropdown;
