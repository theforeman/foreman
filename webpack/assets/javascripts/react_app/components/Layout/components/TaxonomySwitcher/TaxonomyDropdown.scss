@import '~@theforeman/vendor/scss/variables';

.pf-c-masthead .pf-c-toolbar {
  .pf-c-context-selector__menu-search {
    border-bottom: 0;
  }
  .current-taxonomy-v {
    margin-right: 2px;
  }
  .organizations_clear,
  .locations_clear {
    border-bottom: var(--pf-c-context-selector__menu-search--BorderBottomWidth)
      solid var(--pf-c-context-selector__menu-search--BorderBottomColor);
  }
  .pf-c-context-selector__menu-footer > a{
    // to override secondry button text color being white if the button is in a menu
    color: $pf-color-blue-400
  }
}
