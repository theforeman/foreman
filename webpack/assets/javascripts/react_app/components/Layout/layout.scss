@import '~@theforeman/vendor/scss/variables';
@import '../../common/colors.scss';

.secondary-nav-item-pf > .nav-pf-secondary-nav {
  background-color: $nav-pf-vertical-active-bg-color;

  .mobile-active {
    // Mark active Location/Org in Mobile view
    a {
      .list-group-item-value {
        background-color: $nav-pf-vertical-secondary-active-bg-color;
      }
    }
  }

  .list-group-item {
    a {
      background-color: $nav-pf-vertical-active-bg-color;

      &:hover {
        .list-group-item-value {
          background-color: $nav-pf-vertical-secondary-active-bg-color;
        }
      }
    }

    span {
      color: $nav-pf-vertical-active-icon-color;
    }
  }
}

#react-app-root {
  .react-container {
    margin-right: auto;
    margin-left: 200px;
    padding-left: 20px;
    padding-right: 20px;

    &.collapsed-nav {
      margin-left: 75px;
    }
  }

  .nav-pf-vertical {
    background-color: $nav-pf-vertical-secondary-active-bg-color;

    .list-group {
      .list-group-item {
        a {
          span {
            color: $nav-pf-vertical-icon-color;
          }

          &:hover {
            background-color: $nav-pf-vertical-active-bg-color;

            span {
              color: $nav-pf-vertical-active-icon-color;
            }
          }

          &::after {
            color: $nav-pf-vertical-icon-color;
          }
        }

        &:hover {
          a {
            background-color: $nav-pf-vertical-active-bg-color;

            span {
              color: $nav-pf-vertical-active-icon-color;
            }
          }
        }

        &.active {
          a {
            background-color: $nav-pf-vertical-active-bg-color;

            span {
              color: $nav-pf-vertical-active-icon-color;
            }
          }
        }
      }
    }
  }
}

/*
  patternfly-sass hides the vertical-nav on smaller screens, see: https://github.com/patternfly/patternfly-sass/blob/5e7cce3445d5b1af1e16500695b9f33edc6a4f6c/assets/javascripts/patternfly-functions.js#L260
  the react-container needs to be shifted when it happens.
*/
@media only screen and (max-width: $pf-global--breakpoint--md) {
  #react-app-root {
    .react-container {
      margin-left: 0;

      &.collapsed-nav {
        margin-left: 0;
      }
    }
  }
}

#account_menu {
  outline: none;
}

#vertical-spinner {
  margin-top: 15px;
  margin-left: 5px;
}

.navbar-header {
  width: 221.22px;
}

.taxonomy_search {
  max-width: 92%;
  margin: 4px 0 4px 8px;
}
