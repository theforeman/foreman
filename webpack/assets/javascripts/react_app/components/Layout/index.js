import React, { useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { useHistory } from 'react-router-dom';

import {
  ANY_LOCATION_TAXONOMY,
  ANY_ORGANIZATION_TAXONOMY,
} from './LayoutConstants';
import {
  initializeLayout,
  changeActiveMenu,
  changeOrganization,
  changeLocation,
  collapseLayoutMenus,
  expandLayoutMenus,
} from './LayoutActions';
import reducer from './LayoutReducer';
import {
  patternflyMenuItemsSelector,
  selectActiveMenu,
  selectCurrentOrganization,
  selectCurrentLocation,
  selectIsLoading,
  selectIsCollapsed,
} from './LayoutSelectors';
import {
  createInitialTaxonomy,
  combineMenuItems,
  getActiveMenuItem,
} from './LayoutHelper';
import { getIsNavbarCollapsed } from './LayoutSessionStorage';

import Layout from './Layout';

const ConnectedLayout = ({ children, data }) => {
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(
      initializeLayout({
        items: combineMenuItems(data),
        activeMenu: getActiveMenuItem(data.menu).title,
        isCollapsed: getIsNavbarCollapsed(),
        organization:
          data.taxonomies.organizations && data.orgs.current_org
            ? createInitialTaxonomy(
                data.orgs.current_org,
                data.orgs.available_organizations
              )
            : ANY_ORGANIZATION_TAXONOMY,
        location:
          data.taxonomies.locations && data.locations.current_location
            ? createInitialTaxonomy(
                data.locations.current_location,
                data.locations.available_locations
              )
            : ANY_LOCATION_TAXONOMY,
      })
    );
  }, [data, dispatch]);

  const { push: navigate } = useHistory();
  const items = useSelector(state => patternflyMenuItemsSelector(state));
  const isLoading = useSelector(state => selectIsLoading(state));
  const isCollapsed = useSelector(state => selectIsCollapsed(state));
  const activeMenu = useSelector(state => selectActiveMenu(state));
  const currentOrganization = useSelector(state =>
    selectCurrentOrganization(state)
  );
  const currentLocation = useSelector(state => selectCurrentLocation(state));

  return (
    <Layout
      data={data}
      navigate={navigate}
      items={items}
      isLoading={isLoading}
      isCollapsed={isCollapsed}
      activeMenu={activeMenu}
      currentOrganization={currentOrganization}
      currentLocation={currentLocation}
      changeActiveMenu={menu => dispatch(changeActiveMenu(menu))}
      changeOrganization={org => dispatch(changeOrganization(org))}
      changeLocation={loc => dispatch(changeLocation(loc))}
      collapseLayoutMenus={() => dispatch(collapseLayoutMenus())}
      expandLayoutMenus={() => dispatch(expandLayoutMenus())}
    >
      {children}
    </Layout>
  );
};

// export prop-types
export const { propTypes, defaultProps } = Layout;

ConnectedLayout.propTypes = {
  children: propTypes.children,
  data: propTypes.data,
};

ConnectedLayout.defaultProps = {
  children: defaultProps.children,
  data: defaultProps.data,
};

// export reducers
export const reducers = { layout: reducer };

// export connected component
export default ConnectedLayout;
