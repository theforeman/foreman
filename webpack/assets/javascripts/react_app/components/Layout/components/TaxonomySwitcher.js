import React from 'react';
import PropTypes from 'prop-types';
import { Nav, Spinner } from 'patternfly-react';
import { noop } from '../../../common/helpers';

import { locationPropType, organizationPropType } from '../LayoutHelper';
import { ANY_ORGANIZATION_TEXT, ANY_LOCATION_TEXT } from '../LayoutConstants';
import NavItem from './NavItem';
import TaxonomyDropdown from './TaxonomyDropdown';

const TaxonomySwitcher = ({
  currentOrganization,
  currentLocation,
  organizations,
  locations,
  taxonomiesBool,
  isLoading,
  onLocationClick,
  onOrgClick,
}) => (
  <Nav navbar pullLeft className="navbar-iconic">
    {taxonomiesBool.organizations && (
      <TaxonomyDropdown
        taxonomyType="Organization"
        id="organization-dropdown"
        currentTaxonomy={currentOrganization}
        taxonomies={organizations}
        changeTaxonomy={onOrgClick}
        anyTaxonomyText={ANY_ORGANIZATION_TEXT}
        manageTaxonomyText="Manage Organizations"
        anyTaxonomyURL="/organizations/clear"
        manageTaxonomyURL="/organizations"
      />
    )}
    {taxonomiesBool.locations && (
      <TaxonomyDropdown
        taxonomyType="Location"
        id="location-dropdown"
        currentTaxonomy={currentLocation}
        taxonomies={locations}
        changeTaxonomy={onLocationClick}
        anyTaxonomyText={ANY_LOCATION_TEXT}
        manageTaxonomyText="Manage Locations"
        anyTaxonomyURL="/locations/clear"
        manageTaxonomyURL="/locations"
      />
    )}
    {isLoading && (
      <NavItem id="vertical-spinner">
        <Spinner size="md" inverse loading />
      </NavItem>
    )}
  </Nav>
);
TaxonomySwitcher.propTypes = {
  onLocationClick: PropTypes.func,
  onOrgClick: PropTypes.func,
  isLoading: PropTypes.bool,
  currentOrganization: PropTypes.string,
  currentLocation: PropTypes.string,
  organizations: PropTypes.arrayOf(organizationPropType).isRequired,
  locations: PropTypes.arrayOf(locationPropType).isRequired,
  taxonomiesBool: PropTypes.shape({
    locations: PropTypes.bool.isRequired,
    organizations: PropTypes.bool.isRequired,
  }).isRequired,
};
TaxonomySwitcher.defaultProps = {
  isLoading: false,
  currentLocation: ANY_LOCATION_TEXT,
  currentOrganization: ANY_ORGANIZATION_TEXT,
  onLocationClick: noop,
  onOrgClick: noop,
};
export default TaxonomySwitcher;
