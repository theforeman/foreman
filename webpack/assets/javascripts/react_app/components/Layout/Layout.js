import React from 'react';
import PropTypes from 'prop-types';

import { VerticalNav } from 'patternfly-react';
import { noop } from '../../common/helpers';

import { getActive, getCurrentPath, handleMenuClick } from './LayoutHelper';
import TaxonomySwitcher from './components/TaxonomySwitcher';
import UserDropdowns from './components/UserDropdowns';
import './layout.scss';

class Layout extends React.Component {
  componentDidMount() {
    const {
      items,
      data,
      fetchMenuItems,
      changeLocation,
      currentLocation,
      changeOrganization,
      currentOrganization,
      changeActiveMenu,
      activeMenu,
    } = this.props;
    if (items.length === 0) fetchMenuItems(data);

    const activeURLMenu = getActive(data.menu, getCurrentPath());
    if (activeMenu !== activeURLMenu.title) {
      changeActiveMenu(activeURLMenu);
    }

    if (
      data.taxonomies.locations &&
      !!data.locations.current_location &&
      currentLocation !== data.locations.current_location
    ) {
      const initialLocTitle = data.locations.current_location;
      const initialLocId = data.locations.available_locations.find(
        loc => loc.title === initialLocTitle
      ).id;
      changeLocation({ title: initialLocTitle, id: initialLocId });
    }

    if (
      data.taxonomies.organizations &&
      !!data.orgs.current_org &&
      currentOrganization !== data.orgs.current_org
    ) {
      const initialOrgTitle = data.orgs.current_org;
      const initialOrgId = data.orgs.available_organizations.find(
        org => org.title === initialOrgTitle
      ).id;
      changeOrganization({ title: initialOrgTitle, id: initialOrgId });
    }
  }

  render() {
    const {
      items,
      data,
      isLoading,
      changeActiveMenu,
      changeOrganization,
      changeLocation,
      currentOrganization,
      currentLocation,
      activeMenu,
    } = this.props;

    return (
      <VerticalNav
        hoverDelay={100}
        items={items}
        onItemClick={primary =>
          handleMenuClick(primary, activeMenu, changeActiveMenu)
        }
        activePath={`/${activeMenu}/`}
        {...this.props}
      >
        <VerticalNav.Masthead>
          <VerticalNav.Brand
            title={data.brand}
            iconImg={data.logo}
            href={data.root}
          />
          <TaxonomySwitcher
            taxonomiesBool={data.taxonomies}
            currentLocation={currentLocation}
            locations={
              data.taxonomies.locations
                ? data.locations.available_locations
                : []
            }
            onLocationClick={changeLocation}
            currentOrganization={currentOrganization}
            organizations={
              data.taxonomies.organizations
                ? data.orgs.available_organizations
                : []
            }
            onOrgClick={changeOrganization}
            isLoading={isLoading}
          />
          <UserDropdowns
            notificationUrl={data.notification_url}
            user={data.user}
            changeActiveMenu={changeActiveMenu}
          />
        </VerticalNav.Masthead>
      </VerticalNav>
    );
  }
}

Layout.propTypes = {
  currentOrganization: PropTypes.string,
  currentLocation: PropTypes.string,
  isLoading: PropTypes.bool,
  activeMenu: PropTypes.string,
  fetchMenuItems: PropTypes.func,
  changeActiveMenu: PropTypes.func,
  changeOrganization: PropTypes.func,
  changeLocation: PropTypes.func,
  items: PropTypes.arrayOf(
    PropTypes.shape({
      title: PropTypes.string.isRequired,
      className: PropTypes.string,
      iconClass: PropTypes.string.isRequired,
      initialActive: PropTypes.bool,
      subItems: PropTypes.arrayOf(
        PropTypes.shape({
          title: PropTypes.string,
          isDivider: PropTypes.bool,
          className: PropTypes.string,
          href: PropTypes.string.isRequired,
        })
      ),
    })
  ),
  data: PropTypes.shape({
    menu: PropTypes.arrayOf(
      PropTypes.shape({
        type: PropTypes.string.isRequired,
        name: PropTypes.string.isRequired,
        icon: PropTypes.string.isRequired,
        children: PropTypes.any,
      })
    ),
    locations: PropTypes.shape({
      current_location: PropTypes.string,
      available_locations: PropTypes.arrayOf(
        PropTypes.shape({
          href: PropTypes.string.isRequired,
          id: PropTypes.number.isRequired,
          title: PropTypes.string,
        })
      ),
    }),
    orgs: PropTypes.shape({
      current_organization: PropTypes.string,
      available_organizations: PropTypes.arrayOf(
        PropTypes.shape({
          href: PropTypes.string.isRequired,
          id: PropTypes.number.isRequired,
          title: PropTypes.string,
        })
      ),
    }),
    root: PropTypes.string.isRequired,
    logo: PropTypes.string.isRequired,
    notification_url: PropTypes.string.isRequired,
    taxonomies: PropTypes.shape({
      locations: PropTypes.bool.isRequired,
      organizations: PropTypes.bool.isRequired,
    }),
    user: PropTypes.shape({
      current_user: PropTypes.object.isRequired,
      user_dropdown: PropTypes.arrayOf(
        PropTypes.shape({
          children: PropTypes.any,
          icon: PropTypes.string.isRequired,
          name: PropTypes.string.isRequired,
          type: PropTypes.string.isRequired,
        })
      ),
    }),
  }),
};

Layout.defaultProps = {
  items: [],
  data: {},
  currentOrganization: 'Any Organization',
  currentLocation: 'Any Location',
  isLoading: false,
  activeMenu: '',
  fetchMenuItems: noop,
  changeActiveMenu: noop,
  changeOrganization: noop,
  changeLocation: noop,
};

export default Layout;
