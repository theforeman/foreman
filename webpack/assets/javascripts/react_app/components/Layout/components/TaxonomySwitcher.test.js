import { testComponentSnapshotsWithFixtures } from '../../../common/testHelpers';
import TaxonomySwitcher from './TaxonomySwitcher';
import { layoutMock } from '../Layout.fixtures';

const props = {
  organizations: layoutMock.data.orgs.available_organizations,
  taxonomiesBool: layoutMock.data.taxonomies,
  locations: layoutMock.data.locations.available_locations,
  currentLocation: 'location',
  currentOrganization: 'organization',
  isLoading: true,
};

const fixtures = {
  'render TaxonomySwitcher': { ...props },
};

describe('TaxonomySwitcher', () => {
  describe('rendering', () =>
    testComponentSnapshotsWithFixtures(TaxonomySwitcher, fixtures));
});
