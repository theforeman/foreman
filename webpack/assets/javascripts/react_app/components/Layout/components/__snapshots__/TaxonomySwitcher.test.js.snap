// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaxonomySwitcher rendering render TaxonomySwitcher 1`] = `
<Nav
  bsClass="nav"
  className="navbar-iconic"
  justified={false}
  navbar={true}
  pullLeft={true}
  pullRight={false}
  stacked={false}
>
  <TaxonomyDropdown
    anyTaxonomyText="Any Organization"
    anyTaxonomyURL="/organizations/clear"
    changeTaxonomy={[Function]}
    currentTaxonomy="organization"
    id="organization-dropdown"
    manageTaxonomyText="Manage Organizations"
    manageTaxonomyURL="/organizations"
    taxonomies={
      Array [
        Object {
          "href": "/organizations/1-org1/select",
          "id": 1,
          "title": "org1",
        },
        Object {
          "href": "/organizations/2-org2/select",
          "id": 2,
          "title": "org2",
        },
      ]
    }
    taxonomyType="Organization"
  />
  <TaxonomyDropdown
    anyTaxonomyText="Any Location"
    anyTaxonomyURL="/locations/clear"
    changeTaxonomy={[Function]}
    currentTaxonomy="location"
    id="location-dropdown"
    manageTaxonomyText="Manage Locations"
    manageTaxonomyURL="/locations"
    taxonomies={
      Array [
        Object {
          "href": "/locations/1-yaml/select",
          "id": 1,
          "title": "yaml",
        },
        Object {
          "href": "/locations/2-london/select",
          "id": 2,
          "title": "london",
        },
        Object {
          "href": "/locations/3-norway/select",
          "id": 3,
          "title": "norway",
        },
      ]
    }
    taxonomyType="Location"
  />
  <NavItem
    activeHref=""
    activeKey=""
    className=""
    id="vertical-spinner"
  >
    <Spinner
      className=""
      inline={false}
      inverse={true}
      loading={true}
      size="md"
    />
  </NavItem>
</Nav>
`;
